catchup/networkFetcher.go

```diff
@@ -0,0 +1,92 @@
+// License, or (at your option) any later version.
+//
+// go-algorand is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with go-algorand.  If not, see <https://www.gnu.org/licenses/>.
+
+package catchup
+
+import (
+	"context"
+	"errors"
+	"time"
+
+	"github.com/algorand/go-algorand/agreement"
+	"github.com/algorand/go-algorand/config"
+	"github.com/algorand/go-algorand/data/basics"
+	"github.com/algorand/go-algorand/data/bookkeeping"
+	"github.com/algorand/go-algorand/logging"
+	"github.com/algorand/go-algorand/network"
+)
+
+// NetworkFetcher is the interface used to export fetchBlock function from universalFetcher
+type NetworkFetcher interface {
+	// FetchBlock function creates a new universalBlockFetcher and calls the internal fetchBlock method
+	// for a given context and round number
+	FetchBlock(ctx context.Context, round basics.Round) (blk *bookkeeping.Block,
+		cert *agreement.Certificate, downloadDuration time.Duration, err error)
+}
+
+type networkFetcherImpl struct {
+	log          logging.Logger
+	cfg          config.Local
+	net          network.GossipNode
+	peerSelector *peerSelector
+}
+
+// NewNetworkFetcher initializes a NetworkFetcher service
+func NewNetworkFetcher(log logging.Logger, net network.GossipNode, cfg config.Local, pipelineFetch bool) NetworkFetcher {
+	netFetcher := &networkFetcherImpl{
+		net: net,
+		cfg: cfg,
+		log: log,
+	}
+	// creating peerselector for the network fetcher
+	netFetcher.peerSelector = createPeerSelector(net, cfg, pipelineFetch)
+	return netFetcher
+}
+
+// FetchBlock function creates a new universalBlockFetcher and calls the internal fetchBlock method for a given context and round number
+func (netFetcher *networkFetcherImpl) FetchBlock(ctx context.Context, round basics.Round) (blk *bookkeeping.Block,
+	cert *agreement.Certificate, downloadDuration time.Duration, err error) {
+	fetch := makeUniversalBlockFetcher(netFetcher.log, netFetcher.net, netFetcher.cfg)
+	psp, err := netFetcher.peerSelector.getNextPeer()
+	if err != nil {
+		return
+	}
+	peer := psp.Peer
+	httpPeer, ok := peer.(network.HTTPPeer)
+	if !ok {
+		netFetcher.log.Warnf("FetchBlock: non-HTTP peer was provided by the peer selector")
+	}
+	blk, cert, _, err = fetch.fetchBlock(ctx, round, httpPeer)
+	if err != nil {
+		netFetcher.log.Debugf("FetchBlock(%v): Could not fetch: %v", round, err)
+		netFetcher.peerSelector.rankPeer(psp, peerRankDownloadFailed)
+		return
+	}
+	// Check that the block's contents match the block header (necessary with an untrusted block because b.Hash() only hashes the header)
+	if blk == nil || cert == nil {
+		err = errors.New("invalid block download")
+	} else if !blk.ContentsMatchHeader() && blk.Round() > 0 {
```

Comment:
There are three cases you need to consider here.
1. The block download failed (`fetch.fetchBlock` returned error and `ctx` is not the reason). Then you need to rank the peer, as you already do, with `peerRankDownloadFailed`.
2. The block is not consistent (`ContentsMatchHeader` fails). Then you need to rank the peer with `peerRankInvalidDownload`. You also need to see if you need to validated it with `s.auth.Authenticate(block, cert)`.
3. If the block is valid, you need to update the peer selected with the download speed.
```
		peerRank := peerSelector.peerDownloadDurationToRank(psp, blockDownloadDuration)
		peerSelector.rankPeer(psp, peerRank)
```

Reply:
2. The idea was to use `NetworkFetcher` for solely downloading the blocks, and the `importer` plugins defined for indexer can validate the blocks on their side

Reply:
Hmmm. This could be the only way. But will complicate the feedback to the peer selector if the validation fails. 
Then you need a function on `NetworkFetcher` to report to the peer selector `peerRankInvalidDownload`.

Reply:
I'll add the authentication code inside networkFetcher, the alternative- declaring a new function and caching an instance of `psp` to rank the peer seems worse